### PR TITLE
1412 global mutual information

### LINKED ISSUE #1412: add image dissimilarity loss
**Is your feature request related to a problem? Please describe.**
Add image dissimilarity measure between a pair of 1D-3D images for registration.

Like deform loss, image loss is reasonably independent for starting to put these in first. It also makes sense to me to add the three "major" ones in one go: squared-difference, cross-correlation-based, mutual-information. 

Reference code:
https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/loss/image.py

@wyli @kate-sann5100 what do you think?

### PR DESCRIPTION
**Pull request by kate-sann5100**
Fixes #1412 .

### Description
Add global mutual information

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by wyli** [comment]
thanks, this is nice, could you add an integration test? like this file: https://github.com/Project-MONAI/MONAI/blob/master/tests/test_seg_loss_integration.py (could be simpler than it, just wanted to verify that the new losses are differentiable)

**Comment by kate-sann5100** [comment]
> thanks, this is nice, could you add an integration test? like this file: https://github.com/Project-MONAI/MONAI/blob/master/tests/test_seg_loss_integration.py (could be simpler than it, just wanted to verify that the new losses are differentiable)
> 
> * also please merge the latest master into this branch and resolve the conflicts

Sure, I am happy to work on that.

File: monai/losses/deform.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def spatial_gradient(input: torch.Tensor, dim: int) -> torch.Tensor:
         DeepReg (https://github.com/DeepRegNet/DeepReg)
 
     Args:
-        input: the shape should be BCH(WD).
+        pred: the shape should be BCH(WD).
```
Comment (wyli) [inline monai/losses/deform.py:29]:
```suggestion
        x: the shape should be BCH(WD).
```

File: monai/losses/image_dissimilarity.py [hunk 2]
```diff
@@ -76,11 +76,11 @@ def __init__(
             kernel_size: kernel spatial size, must be odd.
             kernel_type: {``"rectangular"``, ``"triangular"``, ``"gaussian"``}. Defaults to ``"rectangular"``.
             reduction: {``"none"``, ``"mean"``, ``"sum"``}
-                Specifies the reduction to apply to the output. Defaults to ``"mean"``.
+                Specifies the reduction to apply to the pred. Defaults to ``"mean"``.
 
                 - ``"none"``: no reduction will be applied.
-                - ``"mean"``: the sum of the output will be divided by the number of elements in the output.
-                - ``"sum"``: the output will be summed.
+                - ``"mean"``: the sum of the pred will be divided by the number of elements in the pred.
```
Comment (wyli) [inline monai/losses/image_dissimilarity.py:82]: pred -> output

File: monai/losses/image_dissimilarity.py [hunk 3]
```diff
@@ -89,7 +89,7 @@ def __init__(
 
         self.ndim = ndim
         if self.ndim not in [1, 2, 3]:
-            raise ValueError(f"Unsupported ndim: {self.ndim}-d, only 1-d, 2-d, and 3-d inputs are supported")
+            raise ValueError(f"Unsupported ndim: {self.ndim}-d, only 1-d, 2-d, and 3-d preds are supported")
```
Comment (wyli) [inline monai/losses/image_dissimilarity.py:92]: preds -> inputs

File: monai/losses/image_dissimilarity.py [hunk 4]
```diff
@@ -157,3 +154,84 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+    ) -> None:
+        """
+        Args:
+            num_bins: number of bins for intensity
+            sigma_ratio: a hyper param for gaussian function
+            reduction: {``"none"``, ``"mean"``, ``"sum"``}
+                Specifies the reduction to apply to the pred. Defaults to ``"mean"``.
+
+                - ``"none"``: no reduction will be applied.
+                - ``"mean"``: the sum of the pred will be divided by the number of elements in the pred.
```
Comment (wyli) [inline monai/losses/image_dissimilarity.py:183]: pred -> output?

**Review by wyli — APPROVED** [review]
thanks, there are a few typos i guess were introduced by auto substitutions, otherwise it looks good to me

File: monai/losses/deform.py (hunk 1 above)
Comment (kate-sann5100) [reply monai/losses/deform.py]: sorry for the type, it is fixed.

File: monai/losses/image_dissimilarity.py (hunk 2 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py]: sorry for the type, it is fixed.

File: monai/losses/image_dissimilarity.py (hunk 3 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py]: sorry for the type, it is fixed.

File: monai/losses/image_dissimilarity.py (hunk 4 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py]: sorry for the type, it is fixed.

File: tests/test_reg_loss_integration.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+            # forward pass
+            output = net(image)
+            loss_input = {"pred": output, "target": target}
+
+            loss_val = loss(**{k: loss_input[k] for k in forward_args})
+
+            # backward pass
+            loss_val.backward()
+            optimizer.step()
```
Comment (wyli) [inline tests/test_reg_loss_integration.py:104]: shall we have some assertions here about the decreasing loss value? or they are omitted on purpose?

Comment (kate-sann5100) [reply tests/test_reg_loss_integration.py:106]: there was a bug in the original implementation of `LocalNormalizedCrossCorrelation` which causes large fluctuations. The bug is fixed now and decreased loss assertion is added.
